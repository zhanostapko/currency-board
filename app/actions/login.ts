"use server";
import { LoginActionType } from "@/types/FormTypes";
import { validateEmail, validatePassword } from "@/utils/validate";

export async function loginAction(
  prevState: LoginActionType,
  formData: FormData
): Promise<LoginActionType> {
  const email = formData.get("email") as string;
  const password = formData.get("password") as string;

  if (!email || !password) {
    return {
      error: "Email and password are required",
      step: "login",
      email,
      otp: null,
    };
  }

  const emailError = validateEmail(email);
  const passwordError = validatePassword(password, "login");

  if (emailError || passwordError) {
    return {
      error: emailError || passwordError,
      step: "login",
      email,
      otp: null,
    };
  }

  if (email === "incorrect@email.com" || password === "incorrect-password") {
    return {
      error: "Invalid email or password",
      step: "login",
      email,
      otp: null,
    };
  }

  const otp = Math.floor(100000 + Math.random() * 900000);

  await new Promise((res) => setTimeout(res, 1500));

  return { error: null, step: "otp", email, otp: otp };
}
