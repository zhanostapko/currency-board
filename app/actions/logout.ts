"use server";
import { cookies } from "next/headers";
import { redirect } from "next/navigation";

export async function logoutAction() {
  const cookieStore = await cookies();
  cookieStore.set("token", "", {
    httpOnly: true,
    secure: process.env.NODE_ENV === "production",
    maxAge: -1,
    path: "/",
  });

  redirect("/login");
}
