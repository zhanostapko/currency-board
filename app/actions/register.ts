"use server";

import { RegisterActionType } from "@/types/FormTypes";
import {
  validateConfirmPassword,
  validateEmail,
  validatePassword,
} from "@/utils/validate";

export async function registerAction(
  prevState: RegisterActionType,
  formData: FormData
): Promise<RegisterActionType> {
  const email = formData.get("email") as string;
  const password = formData.get("password") as string;
  const confirmPassword = formData.get("confirmPassword") as string;

  const emailError = validateEmail(email);
  const passwordError = validatePassword(password, "register");
  const confirmPasswordError = validateConfirmPassword(
    password,
    confirmPassword
  );

  if (emailError || passwordError || confirmPasswordError) {
    return {
      error: emailError || passwordError || confirmPasswordError,
      success: false,
      email,
    };
  }

  await new Promise((res) => setTimeout(res, 1500));

  return { error: null, success: true, email };
}
