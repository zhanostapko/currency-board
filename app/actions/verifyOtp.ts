"use server";
import crypto from "crypto";
import { redirect } from "next/navigation";
import { cookies } from "next/headers";
import { OtpActionType } from "@/types/FormTypes";

export async function verifyOtpAction(
  prevState: OtpActionType,
  formData: FormData
): Promise<OtpActionType> {
  const enteredOtp = formData.get("enteredOtp") as string;
  const serverOtp = formData.get("serverOtp") as string;

  if (!enteredOtp) {
    return { error: "OTP is required", step: "otp" };
  }

  if (enteredOtp !== serverOtp) {
    return { error: "Invalid OTP", step: "otp" };
  }

  await new Promise((res) => setTimeout(res, 1500));

  const token = crypto.randomBytes(32).toString("hex");
  const cookieStore = await cookies();

  cookieStore.set("token", token, {
    httpOnly: true,
    secure: process.env.NODE_ENV === "production",
    maxAge: 3600,
    path: "/",
  });

  redirect("/dashboard");
}
