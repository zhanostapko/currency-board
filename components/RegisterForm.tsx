"use client";

import { useActionState, useEffect } from "react";
import { useRouter } from "next/navigation";
import Input from "@/components/Shared/Input";
import SubmitButton from "./SubmitButton";
import { registerAction } from "@/app/actions/register";

import Link from "next/link";
import AuthFormWrapper from "./AuthWrapper";

const RegisterForm = () => {
  const [state, formAction] = useActionState(registerAction, {
    error: null,
    success: false,
    email: "",
  });
  const router = useRouter();

  useEffect(() => {
    if (state.success) {
      setTimeout(() => router.push("/login"), 2000);
    }
  }, [state.success, router]);

  return (
    <AuthFormWrapper
      title={`${state.success ? "" : "Create an Account"}`}
      subtitle={`${state.success ? "" : "Sign up to get started"}`}
    >
      {state.success ? (
        <p className="text-green-700 font-medium text-sm bg-green-100 p-3 rounded-lg border border-green-400">
          ðŸŽ‰ Your account has been successfully created! Redirecting to Sign
          In...
        </p>
      ) : (
        <form
          data-testid="register-form"
          className="flex flex-col gap-4"
          action={formAction}
        >
          <Input
            id="email"
            label="Email"
            type="email"
            name="email"
            required
            defaultValue={state.email}
          />
          <Input
            id="password"
            label="Password"
            type="password"
            name="password"
            required
          />
          <Input
            id="confirmPassword"
            label="Confirm Password"
            type="password"
            name="confirmPassword"
            required
          />
          {state.error && <p className="text-red-500 text-xs">{state.error}</p>}
          <SubmitButton title="Sign up" />
        </form>
      )}
      <Link href="/login">
        <p className="text-gray-500 mt-2 text-sm cursor-pointer hover:underline">
          {`${
            state.success
              ? "Go to sign in page"
              : "Already have an account? Login"
          }`}
        </p>
      </Link>
    </AuthFormWrapper>
  );
};

export default RegisterForm;
