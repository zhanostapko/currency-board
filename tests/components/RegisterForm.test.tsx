import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import RegisterForm from "../../components/RegisterForm";
import { useRouter } from "next/navigation";
import { registerAction } from "../../app/actions/register";
import { act, useActionState } from "react";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("../../app/actions/register", () => ({
  registerAction: jest.fn(),
}));

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useActionState: jest.fn(),
}));

describe("RegisterForm Component", () => {
  const mockRouter = { push: jest.fn() };
  beforeEach(() => {
    jest.clearAllMocks();
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
    (useActionState as jest.Mock).mockReturnValue([
      { success: false, error: null, email: "" },
      jest.fn(),
    ]);
  });

  test("renders registration form", () => {
    const { asFragment } = render(<RegisterForm />);

    expect(screen.getByLabelText("Email")).toBeInTheDocument();
    expect(screen.getByLabelText("Password")).toBeInTheDocument();
    expect(screen.getByLabelText("Confirm Password")).toBeInTheDocument();
    expect(screen.getByText("Sign up")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("shows error when passwords do not match", async () => {
    const { asFragment } = render(<RegisterForm />);

    fireEvent.change(screen.getByLabelText("Email"), {
      target: { value: "test@example.com" },
    });
    fireEvent.change(screen.getByLabelText("Password"), {
      target: { value: "password123" },
    });
    fireEvent.change(screen.getByLabelText("Confirm Password"), {
      target: { value: "password321" },
    });

    fireEvent.click(screen.getByText("Sign up"));

    (useActionState as jest.Mock).mockReturnValueOnce([
      {
        success: false,
        error: "Passwords do not match",
        email: "test@example.com",
      },
      jest.fn(),
    ]);

    act(() => {
      render(<RegisterForm />);
    });

    await waitFor(() => {
      expect(screen.getByText("Passwords do not match")).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test("calls registerAction on valid form submission", async () => {
    (registerAction as jest.Mock).mockResolvedValue({ success: true });

    const { asFragment } = render(<RegisterForm />);

    fireEvent.change(screen.getByLabelText(/email/i), {
      target: { value: "test@example.com" },
    });

    const [passwordInput, confirmPasswordInput] =
      screen.getAllByLabelText(/password/i);

    fireEvent.change(passwordInput, { target: { value: "password123" } });
    fireEvent.change(confirmPasswordInput, {
      target: { value: "password123" },
    });

    const button = screen.getByRole("button", { name: /sign up/i });
    fireEvent.click(button);

    (useActionState as jest.Mock).mockReturnValueOnce([
      { success: true, error: null, email: "test@example.com" },
      jest.fn(),
    ]);

    act(() => {
      render(<RegisterForm />);
    });

    await waitFor(() => {
      expect(
        screen.getAllByText(
          /Your account has been successfully created! Redirecting to Sign In/i
        ).length
      ).toBeGreaterThan(0);
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
