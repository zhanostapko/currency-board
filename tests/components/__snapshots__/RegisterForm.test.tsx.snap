// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm Component calls registerAction on valid form submission 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center "
  >
    <div
      class="bg-white p-10 rounded-xl shadow-xl w-96 text-center"
    >
      <h2
        class="text-3xl font-bold text-gray-900 mb-2"
      >
        Create an Account
      </h2>
      <p
        class="text-gray-600 mb-6"
      >
        Sign up to get started
      </p>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="flex flex-col gap-4"
        data-testid="register-form"
      >
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="email"
        >
          Email
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="password"
          name="password"
          required=""
          type="password"
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="confirmPassword"
        >
          Confirm Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="confirmPassword"
          name="confirmPassword"
          required=""
          type="password"
        />
        <button
          class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
          type="submit"
        >
          Sign up
        </button>
      </form>
      <a
        href="/login"
      >
        <p
          class="text-gray-500 mt-2 text-sm cursor-pointer hover:underline"
        >
          Already have an account? Login
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RegisterForm Component renders registration form 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center "
  >
    <div
      class="bg-white p-10 rounded-xl shadow-xl w-96 text-center"
    >
      <h2
        class="text-3xl font-bold text-gray-900 mb-2"
      >
        Create an Account
      </h2>
      <p
        class="text-gray-600 mb-6"
      >
        Sign up to get started
      </p>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="flex flex-col gap-4"
        data-testid="register-form"
      >
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="email"
        >
          Email
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="password"
          name="password"
          required=""
          type="password"
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="confirmPassword"
        >
          Confirm Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="confirmPassword"
          name="confirmPassword"
          required=""
          type="password"
        />
        <button
          class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
          type="submit"
        >
          Sign up
        </button>
      </form>
      <a
        href="/login"
      >
        <p
          class="text-gray-500 mt-2 text-sm cursor-pointer hover:underline"
        >
          Already have an account? Login
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RegisterForm Component shows error when passwords do not match 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center "
  >
    <div
      class="bg-white p-10 rounded-xl shadow-xl w-96 text-center"
    >
      <h2
        class="text-3xl font-bold text-gray-900 mb-2"
      >
        Create an Account
      </h2>
      <p
        class="text-gray-600 mb-6"
      >
        Sign up to get started
      </p>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="flex flex-col gap-4"
        data-testid="register-form"
      >
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="email"
        >
          Email
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="password"
          name="password"
          required=""
          type="password"
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="confirmPassword"
        >
          Confirm Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="confirmPassword"
          name="confirmPassword"
          required=""
          type="password"
        />
        <button
          class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
          type="submit"
        >
          Sign up
        </button>
      </form>
      <a
        href="/login"
      >
        <p
          class="text-gray-500 mt-2 text-sm cursor-pointer hover:underline"
        >
          Already have an account? Login
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
